File: src/master-data-reconciliation/src/main/java/io/ioka/demo/reconcile/ProviderFixture.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.demo.reconcile;
import com.sun.net.httpserver.HttpServer;
import io.ioka.munarium.client.model.Json;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.atomic.*;
import java.util.regex.Pattern;

/** A canned protocol responder; no model runtime, input mount, or private oracle. */
public final class ProviderFixture {
    private ProviderFixture() {}
    public static void run() throws Exception {
        var server=HttpServer.create(new InetSocketAddress(11434),0);var fail=new AtomicBoolean();var calls=new AtomicInteger();
        server.createContext("/",exchange->{
            int status=200;Object response;
            try {
                String path=exchange.getRequestURI().getPath();
                if(path.equals("/api/tags")) response=Map.of("models",List.of(Map.of("name","reconcile-selected","model","reconcile-selected")));
                else if(path.equals("/calls")) response=Map.of("calls",calls.get());
                else if(path.equals("/fail")) {fail.set(true);response=Map.of("ok",true);}
                else if(path.equals("/reset")) {fail.set(false);response=Map.of("ok",true);}
                else if(path.equals("/api/chat")) {
                    calls.incrementAndGet();
                    if(fail.get()) {status=503;response=Map.of("error","controlled provider outage");}
                    else {
                        var body=Json.MAPPER.readTree(exchange.getRequestBody());StringBuilder prompt=new StringBuilder();for(var message:body.path("messages")) prompt.append(message.path("content").asText()).append('\n');
                        String evidence=prompt.toString().split("EVIDENCE_START")[1].split("EVIDENCE_END")[0];
                        var citation=Pattern.compile("\\[([^\\[\\]\\s]+/[^\\[\\]\\s]+)\\]").matcher(evidence);var recommendation=Pattern.compile("Recommendation: ([a-z_]+)").matcher(evidence);
                        if(!citation.find() || !recommendation.find()) throw new IllegalStateException("No policy evidence");
                        String answer=FilesUtil.json(Map.of("explanation",evidence.substring(evidence.indexOf("Policy revision:")).replaceAll("\\[[^]]+\\]","").trim(),"recommendation",recommendation.group(1),"disposition","review_required","citations",List.of(citation.group(1))));
                        response=Map.of("model",body.path("model").asText(),"done",true,"done_reason","stop","message",Map.of("role","assistant","content",answer),"prompt_eval_count",24,"eval_count",32);
                    }
                } else {status=404;response=Map.of("error","unknown route");}
                byte[] bytes=FilesUtil.json(response).getBytes(StandardCharsets.UTF_8);exchange.getResponseHeaders().set("Content-Type","application/json");exchange.sendResponseHeaders(status,bytes.length);exchange.getResponseBody().write(bytes);
            } catch(Exception error) {exchange.sendResponseHeaders(500,-1);} finally {exchange.close();}
        });server.start();System.out.println("Controlled stewardship fixture; no model inference.");
    }
}

File: src/employee-policy-assistant/Policy.Harness/ProviderFixture.cs
// SPDX-License-Identifier: Apache-2.0
using System.Net;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Policy.Harness;

/// <summary>Controlled Ollama protocol fixture; builds responses only from the request, with no oracle mount.</summary>
public static class ProviderFixture
{
    public static async Task Run()
    {
        var listener = new HttpListener(); listener.Prefixes.Add("http://+:11434/"); listener.Start();
        var calls = new List<JsonElement>();
        var fail = false;
        while (true)
        {
            var context = await listener.GetContextAsync();
            object response;
            var path = context.Request.Url!.AbsolutePath;
            if (path == "/api/tags") response = new { models = new[] { new { name = "policy-selected", model = "policy-selected" } } };
            else if (path == "/calls") response = calls;
            else if (path == "/fail") { fail = true; response = new { ok = true }; }
            else if (path == "/reset") { fail = false; response = new { ok = true }; }
            else if (path == "/api/chat")
            {
                using var request = await JsonDocument.ParseAsync(context.Request.InputStream);
                var body = request.RootElement; calls.Add(body.Clone());
                if (fail) { context.Response.StatusCode = 503; response = new { error = "controlled provider outage" }; }
                else
                {
                    var prompt = string.Join("\n", body.GetProperty("messages").EnumerateArray().Select(m => m.GetProperty("content").GetString()));
                    var answer = "[\"policy\",\"procedure\"]";
                    if (prompt.Contains("EVIDENCE_START"))
                    {
                        var question = prompt.Split("QUESTION_START").Last().Split("QUESTION_END")[0].ToLowerInvariant();
                        var evidence = prompt.Split("EVIDENCE_START").Last().Split("EVIDENCE_END")[0];
                        var topic = new[] { "moonbase", "retention", "commute", "training", "equipment", "leave" }.FirstOrDefault(question.Contains) ?? "scope";
                        var blocks = Regex.Matches(evidence, @"\[([^\[\]\s]+/[^\[\]\s]+)\]([^\[]*)", RegexOptions.Singleline);
                        var selected = blocks.Cast<Match>().FirstOrDefault(m => m.Groups[2].Value.Contains("TOPIC:" + topic));
                        var fallback = blocks.Cast<Match>().FirstOrDefault();
                        answer = selected is null ? "The requested answer is not available in the supplied documents. " + (fallback?.Groups[1].Value is string label ? "[" + label + "]" : "") : selected.Groups[2].Value.Split("TOPIC:" + topic).Last().Trim() + " [" + selected.Groups[1].Value + "]";
                    }
                    response = new { model = body.GetProperty("model").GetString(), done = true, done_reason = "stop", message = new { role = "assistant", content = answer }, prompt_eval_count = 12, eval_count = 8 };
                }
            }
            else { context.Response.StatusCode = 404; response = new { error = "unknown fixture route" }; }
            var bytes = JsonSerializer.SerializeToUtf8Bytes(response);
            context.Response.ContentType = "application/json"; context.Response.ContentLength64 = bytes.Length;
            await context.Response.OutputStream.WriteAsync(bytes); context.Response.Close();
        }
    }
}
